From 7842259d11ad14b083a3097c7348a8cab77d72c6 Mon Sep 17 00:00:00 2001
From: Athanasios Oikonomou <athoik@gmail.com>
Date: Sun, 29 Sep 2013 16:36:01 +0300
Subject: [PATCH] Handle DTS Passthrough and Downmix

When downmix is enabled we remove on runtime the dts caps to give priority on dtsdownmix.

Also provide a simple way to check if ac3plus is enabled using cheap access call

diff --git a/src/gstdvbaudiosink.c b/src/gstdvbaudiosink.c
index ac1c145..9a201d6 100644
--- a/src/gstdvbaudiosink.c
+++ b/src/gstdvbaudiosink.c
@@ -62,6 +62,7 @@
 #endif
 #include <unistd.h>
 #include <stdint.h>
+#include <stdio.h>
 #include <string.h>
 #include <sys/ioctl.h>
 #include <sys/socket.h>
@@ -301,24 +302,30 @@ gst_dvbaudiosink_base_init (gpointer klass)
 	gst_element_class_set_details (element_class, &element_details);
 }
 
+static gboolean get_downmix_setting()
+{
+	FILE *f;
+	char buffer[32] = {0};
+	f = fopen("/proc/stb/audio/ac3", "r");
+	if (f)
+	{
+		fread(buffer, sizeof(buffer), 1, f);
+		fclose(f);
+	}
+	return !strncmp(buffer, "downmix", 7);
+}
+
+static gboolean get_ac3plus_support()
+{
+	return access("/proc/stb/audio/ac3plus", R_OK) == 0;
+}
+
 static GstCaps *
 gst_dvbaudiosink_get_caps (GstBaseSink *basesink)
 {
 //	GstDVBAudioSink *self = GST_DVBAUDIOSINK (basesink);
 //	gchar *strcaps;
 	GstCaps *caps;
-	static int eac3_support;
-
-	if (!eac3_support) {
-		int fd = open("/proc/stb/audio/ac3plus", O_RDONLY);
-		if (fd >= 0) {
-			eac3_support = 1;
-			close(fd);
-		}
-		else
-			eac3_support = -1;
-	}
-
 	int eac3_pos = 12;
 
 	if (hwtype == DM8000) {
@@ -327,13 +334,16 @@ gst_dvbaudiosink_get_caps (GstBaseSink *basesink)
 		--eac3_pos;
 	}
 	else
-		caps = gst_static_caps_get(&hwtemplate->static_caps);
+		caps = gst_caps_copy(&hwtemplate->static_caps.caps);
 
-	if (eac3_support < 0) {
-		if (eac3_pos == 12)
-			caps = gst_caps_copy(&hwtemplate->static_caps.caps);
+	if (!get_ac3plus_support()) {
 		gst_caps_remove_structure(caps, eac3_pos); // remove x-eac3
 		gst_caps_remove_structure(caps, eac3_pos); // remove x-private-eac3
+        }
+
+	if (hwtemplate == &sink_factory_broadcom_dts && get_downmix_setting()) {
+		gst_caps_remove_structure(caps, 9); // remove x-private1-dts
+		gst_caps_remove_structure(caps, 8); // remove x-dts
 	}
 
 //	strcaps = gst_caps_to_string(caps);
-- 
1.7.10.4

